gateway/ln-gateway/src/state_machine/pay.rs

```diff
@@ -182,20 +182,104 @@ impl GatewayPayInvoice {
+        match Self::await_get_payment_parameters(
+            global_context,
+            pay_invoice_payload.contract_id,
+            context.clone(),
+        )
+        .await
+        {
+            Ok((contract, payment_parameters)) => {
+                Self::buy_preimage(
                     context.clone(),
-                    result,
+                    contract.clone(),
+                    payment_parameters.clone(),
                     common.clone(),
-                    payload.clone(),
-                ))
-            },
-        )]
+                    pay_invoice_payload.clone(),
+                )
+                .await
+            }
+            Err(e) => {
+                warn!("Failed to get payment parameters: {e:?}");
+                match e.contract.clone() {
+                    Some(contract) => GatewayPayStateMachine {
+                        common,
+                        state: GatewayPayStates::CancelContract(Box::new(
+                            GatewayPayCancelContract { contract, error: e },
+                        )),
+                    },
+                    None => GatewayPayStateMachine {
+                        common,
+                        state: GatewayPayStates::OfferDoesNotExist(e.contract_id),
+                    },
+                }
+            }
+        }
+    }
+
+    async fn buy_preimage(
+        context: GatewayClientContext,
+        contract: OutgoingContractAccount,
+        payment_parameters: PaymentParameters,
+        common: GatewayPayCommon,
+        payload: PayInvoicePayload,
+    ) -> GatewayPayStateMachine {
+        debug!("Buying preimage contract {contract:?}");
+        // Verify that this client is authorized to receive the preimage.
+        if let Err(err) = Self::verify_preimage_authentication(
+            &context,
+            payload.payment_hash,
+            payload.preimage_auth,
+            contract.clone(),
+        )
+        .await
+        {
+            warn!("Preimage authentication failed: {err} for contract {contract:?}");
+            return GatewayPayStateMachine {
+                common,
+                state: GatewayPayStates::CancelContract(Box::new(GatewayPayCancelContract {
+                    contract,
+                    error: err,
+                })),
+            };
+        }
+
+        if let Some(client) =
+            Self::check_swap_to_federation(context.clone(), payment_parameters.invoice.clone())
+                .await
+        {
+            Self::buy_preimage_via_direct_swap(
+                client,
+                payment_parameters.invoice.clone(),
+                contract.clone(),
+                common.clone(),
+            )
+            .await
+        } else {
+            Self::buy_preimage_over_lightning(
+                context,
+                payment_parameters,
+                contract.clone(),
+                common.clone(),
+            )
+            .await
+        }
```

Comment:
Buying the preimage in the trigger future is the correct way to go imo, but we need to consider cancel safety: are we sure that if the trigger future gets interrupted and re-run no errors are thrown that would make the state machine fail/refund while the payment was actually sent?

Reply:
I'll try to add a test to cover this scenario

Reply:
I've tested by changing the code to call  `fetch_parameters_and_pay` twice instead of once. And all the tests and load tests works. I'll create an issue to try to convert this scenario into a proper automated test.